### ISSUE TITLE
Annotation for fields that we don't want to be compared

### ISSUE THREAD

**Issue by ralkan** [google/built_value.dart#269]
I've had an issue with using flutter_redux and when creating a ViewModel for the store to pass on to the widget, I wasn't able to compare the old ViewModel instance with the new one every time a change happened. Since I had to pass a `String code` and a `Function onChange` (for dispatching actions) field in the ViewModel.

And when the field `code` is being compared it will return `true` however because the `onChange` field is a closure function, it always compares as `false` so the widget gets updated even when there is no direct change in the state.

```
(store) => new RoomTypeDropdownViewModel((b) => b
      ..code = store.state.roomTypeCode
      ..onChanged = (v) => store.dispatch({
        'type': SearchRoomTypeActions.Select,
        'room_type': v
      }));
```

I think we need a way to make `bool operator ==(other)` ignore specific fields.

**Comment by davidmorgan** [comment]
Thanks for filing and for the PR :)

In your example, I guess it would work to create the function once rather than inline? I guess the main problem is that this is awkward to write? i.e. you'd have to hoist the function out to a top level/static function.

High level comments on the PR:

 1) There's an automated note about signing the Contributor License Agreement; please take a look. This should be straightforward to do. Let me know if you have any problems :)
 2) You'll need to ignore the same fields for hashCode, too.
 3) I agree about `ignored` being awkward. How would you feel about inverting the logic and having a `keyField` annotation instead; if the annotation is present on any fields, then only `keyField`s are used for `hashCode`/`operator==`. (Alternative name ideas: `compareBy`, `compareUsing`.)

Thanks!

**Comment by ralkan** [comment]
Thanks for reopening the issue and your fast response! :)

In my example, unfortunately there seems to be no way (maybe I'm missing something) to statically define the function inside the class since the `store` value that is used inside the inline function comes from a parent scope. So I need a way to dynamically create a curry-like function every time.

Responses to your high level comments:

3 - IMHO, I think that inverting the logic would be quite cumbersome to use and would defeat the purpose of `source_gen` & `built_value` to explicitly annotate all the `keyFields` that you want to be compared (which I think is the case most of the time). :) What do you think about `ignoreComparison`, `incomparable` or `noCompare`?

1 - About that, I actually did. Just before opening this PR. But it seems that my username does indeed match with the one I provided in the CLA but unfortunately my primary e-mail address here on GitHub did not match the one I used to sign the CLA with. After that I went and updated my primary GitHub e-mail to match it, but the CLA check needs to be manually triggered again for a recheck I think.

2 - Thanks for reminding me, I will update the code to include this.

Kindest Regards,

**Comment by ralkan** [comment]
I updated my commit author and force pushed the commit again. I think that that's what triggered it. Not sure. :)
Yes, I don't like it that way either, it seems like too much work for so little gain.
And, then again the `Map` created from store needs to be comparable too and since it has to contain a `Function` for `dispatch` in a key, this again would not be comparable, right?

I really like the suggestion!

Your suggestion made me think of optionally supporting `nullable` and `memoized` in `@BuilValueField` too but I think I'll leave them be as stand-alone strings for usage convenience, what's your opinion on this approach, do you think that I should include them?

**Comment by davidmorgan** [comment]
Good question.

Let's just have `compare` for now, can look at adding/migrating the others later.

**Comment by ralkan** [comment]
Great! 👍

**Comment by davidmorgan** [comment]
Published as 4.4.0. Thanks!

**Comment by ralkan** [comment]
Thanks! @davidmorgan
